pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java

```diff
@@ -424,6 +424,23 @@ public byte[] getSchemaVersion() {
         }
     }
 
+    @Override
+    public byte[] getSchemaId() {
+        if (msgMetadata.hasSchemaId()) {
+            byte[] schemaId = msgMetadata.getSchemaId();
+            return (schemaId.length == 0) ? null : schemaId;
+        }
+        return null;
+    }
+
+    private byte[] getSchemaFlag() {
+        if (msgMetadata.hasSchemaId()) {
+            byte[] schemaId = msgMetadata.getSchemaId();
+            return (schemaId.length == 0) ? null : schemaId;
+        }
```

Comment:
Now if the message has schema id, the schema id will be passed to all places that need a schema version. This should be a wrong behavior because these components only retrieve the schema from **built-in** schema registry. Passing the schema id from external schema registry makes no sense.

Reply:
Yes, you are right, the schema ID should only be used to decode the message payload.

- Changed get schemaInfo to use the schema version.
- Changed get schema reader to use the schema version.

Reply:
I still think we should not mix schema id and version even in decoding. For example, you changed

```java
            return decode(schema.supportSchemaVersioning() ? getSchemaIVersion() : null);
```

to

```java
            return decode(schema.supportSchemaVersioning() ? getSchemaIdOrVersion() : null);
```

It changes the semantics of the `supportSchemaVersioning` method that if it's true, schema id could be used to decode the payload, which is unexpected.

Why not just keep the original logic not changed and add a separated method to decode value by schema id?

```java
    @Override
    public T getValue() {
        if (msgMetadata.hasSchemaId()) {
            return decodeBySchemaId(msgMetadata.getSchemaId());
        }
        // ...
```